[1.10] GuiTabs for survival inventory, patch to solve inventory tab conflicts and reduce keybind usage.

AbsolemJackdaw opened the pull request:
Intended to solve conflicts between mods that add inventories for the player and make a tab for them in the survival inventory. also allows for less key-bind usage.

added ~~survival~~ GuiTabs that do not need to be registered. Instantiating them is enough.

added GuiTablist class, an encapsulating class that handles all the drawing, button, and page logic to keep the code in GuiContainer minimal.

patched GuiInventory to call the super for actionPerformed, so it will pick up the button's from the guitabs

patched GuiCntainer with a few simple lines that refer to the SurvivalTabsList class, a class that handles all logic and drawing for the survival tabs.

_Only works for GuiContainers._

AbsolemJackdaw committed:
Intended to solve conflicts between mods that add inventories for the player and make a tab for them in the survival inventory. also allows for less key-bind usage.

added survival tabs that can be registered instantiating them.
registers their individual id depending on the container they have been attached too.
Only works for GuiContainers.

added SurvivalTablist class, an encasing class that handles all the drawing, button, and page logic to keep the code in GuiContainer minimal.

patched GuiInventory to call the super for actionPerformed, so it will pick up the button's from the survival tabs

patched guicontainer with a few simple lines that refer to the SurvivalTabsList class, a class that handles all logic and drawing for the survival tabs.

gitignore

HenryLoenwind commented:
This might need a test mod to show what it's doing and some docs on how to use it. From your description I think I may like it.

AbsolemJackdaw commented:
I will make sure to add to the wiki ! closing of with some images and it will be available in a few minutes.
I made sure it was fairly easy to use

AbsolemJackdaw commented:
wiki now has a short tutorial and some images and gif to show how stuff works.

https://github.com/ArtixAllMighty/MinecraftForge/wiki

AbsolemJackdaw committed:
fixed resource location not being drawn correctly.
removed item zlevel (unused / not useful)
fixed typo for 'draw' in SurvivalTabList

fixed a reference issue

shadowfacts commented on patches/minecraft/net/minecraft/client/gui/inventory/GuiContainer.java.patch:
```diff
@@ -1,6 +1,45 @@
+     private ItemStack field_146994_N;
+-
++    private net.minecraftforge.client.player.inventory.tabs.SurvivalTabList survivalTabs;
++    
```
No whitespace changes.

AbsolemJackdaw replied:
Sorry, but my editor doesn't show any added whitespace, and genPatches results in no new patch, so I havent changed anything to it since I first edited it.

What do I do ?

HenryLoenwind replied:
Line 9 contains 4 spaces. Remove them, then lines 7 and 9 should go away.

AbsolemJackdaw replied:
I'll figure this out later .... i have no spaces in my editor whatsoever :s

shadowfacts commented on src/main/java/net/minecraftforge/client/player/inventory/tabs/SurvivalTabList.java:
```diff
@@ -0,0 +1,168 @@
+package net.minecraftforge.client.player.inventory.tabs;
```
Missing license header.

AbsolemJackdaw replied:
Licence header added

shadowfacts commented on src/main/java/net/minecraftforge/client/player/inventory/tabs/SurvivalTabList.java:
```diff
@@ -0,0 +1,168 @@
+
+    private boolean renderSurvivalInventoryHoveringText(SurvivalTabs tab, GuiContainer screen, int mouseX, int mouseY, FontRenderer fontRendererObj)
+    {
+        net.minecraftforge.fml.client.config.GuiUtils.drawHoveringText(Arrays.<String> asList(new String[] { I18n.format(tab.getLocalizedName()) }), mouseX,
```
This isn't a patch so you don't have to use fully-qualified names.

shadowfacts commented on src/main/java/net/minecraftforge/client/player/inventory/tabs/SurvivalTabs.java:
```diff
@@ -0,0 +1,218 @@
+package net.minecraftforge.client.player.inventory.tabs;
```
Missing license header

AbsolemJackdaw replied:
Licence header added

TehNut commented on src/main/java/net/minecraftforge/client/player/inventory/tabs/SurvivalTabList.java:
```diff
@@ -0,0 +1,168 @@
+import net.minecraft.util.ResourceLocation;
+
+/**
+ * @author Subaraki
```
Don't add author tags.

TehNut commented on src/main/java/net/minecraftforge/client/player/inventory/tabs/SurvivalTabs.java:
```diff
@@ -0,0 +1,218 @@
+import net.minecraft.util.ResourceLocation;
+
+/**
+ * @author Subaraki
```
Don't add author tags.

AbsolemJackdaw committed:
added licence headers.
removed fully qualified name.

removed author tags

HenryLoenwind commented:
So, just to make sure I understand this correctly:

This makes it so one can add tabs to any GUI (provided you can reference its class). The original GUI is decorated with tab headers by your code, and when one is clicked one's code takes over and can do whatever it wants...but it is expected to open a new GUI. And there's a way to tell your code that that new GUI would like to have the same tabs as any other GUI so it looks for the user as if it was one GUI. --- Neat.

Now, what about GUIs that need a container (because they are GUIs for some block/machine), like the chest GUI you show in your GIF? I guess adding container-less tabs to such a GUI would be easy but then switching tabs back to the container-GUI would be impossible, as would be adding a container GUI as tab to something else?

The case I'd like to see would be to add the player inventory GUI (and all its tab) to chests. This is something I'd be considering doing to my machines.

Another case would be to add a tab to the player inventory that only shows up if the player has a special item in their inventory (and if there are multiple of this item, then also add multiple tabs). This is something I'd love to do for the Ender IO Inventory Panel Remote.

(Sorry if this reads like a wish list. It is ;)))

AbsolemJackdaw commented:
If the machines are yours, there should be no problem. If they use nbt to store things in the tile entity, things might get more complicated.

the reason i could attach the workbench (or i could even have attached an ender chest) is because there is no specific tile entity data needed.

the very first utility of this patch thingy is to add an inventory to the player without having to rebind keys, or tracking down other mods that add tabs on their own. (i have had to do both in the past.)

-~~If you want to only add the tabs to your chest, make a tab for the chest, but don't append it to the player inventory. (new SurvivalTabs("crafting_gui", new ItemStack(Blocks.CRAFTING_TABLE), GuiInventory.class) < tab is appended to guiInventory and will be drawn when the guiinventory is opened). instead : 
new SurvivalTabs("crafting_gui", new ItemStack(Blocks.CRAFTING_TABLE), GuiChest.class)
and just call the createcontinuity for chests and the player inventory.
this way, all player tabs are copied to the chest gui, plus your tab, that will not show up on the player. opening the player gui will not show the tab, opening a chest will show all tabs.~~

_Messed around with chests a bit. it is possible to get the block's location by saving it to the player in a PlayerInteractEvent and using that BlockPos to open the gui in the packet. There's an inconsistency however between chest and player inventory, as the player inventory doesn't have the inventory chest.
Let me tinker a bit more ..._

**The Best way to approach this is to add the chest tab to the inventory, and have a check to send a packet or not. the check can consist of a saved blockpos that will turn to null when you feel like it. (closing gui, etc) I don't think it's a thing to be solved by me, but rather rely on the inventivity of the mod author. but it IS possible.**

the item thingy is a nice idea. i haven't thought about that. That is something we could brainstorm about. I have an irc channel called SubHaven in which we could talk in depth about this, if you'd like to.
**same goes for the item. the tab will have to exist, but can only open when you have the item in the inventory. the tab system relies to much on the ID's for drawing and logic to be able to inject another tab without having gaps. However, I did make the list of tabs available (commit to come), so if mod devs would want to add tabs to the list in some hacky way, they can.**

AbsolemJackdaw committed:
added more header

fixed whitespace.
moved init line to field decleration to keep clean and gain space

AbsolemJackdaw committed:
removed redundant id's for tabs.
made tabs available from a gui container with a getter.
replaced all occurences of the field for tabs in survivaltablist with getter.
refactored some names to have more accurate names.
the guicontainer's list of available tabs is now a copy of the registered tabs so it has no consequence adding or modifying it when a gui is opened. (paired with removed tab id's this could allow adding tabs when having items in the inventory, or the like)

renamed SurvivalTabs and SurvivalTabList to more appropriate names;
renamed survivaltabs field in GuiContainer to guiTabs for more appropriate naming.
made that field protected.

AbsolemJackdaw commented:
The latest commit allows for a more flexible manipulation of the gui tab list.

GotoLink commented on src/main/java/net/minecraftforge/client/player/inventory/tabs/GuiTabList.java:
```diff
@@ -0,0 +1,193 @@
+				int index = tab.getTabIndex();
+				if (index >= start && index < end)
+				{
+					if (tab != null && this.isMouseOverTab(tab, x, y, xSize, ySize))
```
tab can't be null at that point.

GotoLink commented on src/main/java/net/minecraftforge/client/player/inventory/tabs/GuiTabList.java:
```diff
@@ -0,0 +1,193 @@
+		maxPages = ((getTabs().size() - 12) / 12) + 1;
+
+		for (GuiTab tab : getTabs())//set the first page
+			if (container.getClass().equals(tab.getTargetGui()))
```
Try only calling container.getClass() once ?
I think since you create one list for each gui, better be efficient at creation time.

GotoLink commented on src/main/java/net/minecraftforge/client/player/inventory/tabs/GuiTabList.java:
```diff
@@ -0,0 +1,193 @@
+	{
+		this.container = container;
+		tabs = new ArrayList<GuiTab>(GuiTab.getTabsForGui(container.getClass())); //get a copy of the stored list so it can be modified without consequence
+		maxPages = ((getTabs().size() - 12) / 12) + 1;
```
Make maxPages final, and have that 12 in a single place instead of everywhere ?

AbsolemJackdaw replied:
maximum tabs per page, and maxPages isn't the same.
But yes, i'll make a final field for the maxTabsPerPage

GotoLink commented on src/main/java/net/minecraftforge/client/player/inventory/tabs/GuiTabList.java:
```diff
@@ -0,0 +1,193 @@
+					tab.draw(container, isFront, guiLeft, guiTop, xSize, ySize, itemRender, fontRendererObj);
+				}
+
+				if (tab.isActiveTab(container.getClass()) && isFront)
```
If you don't want to use a bit operation for those cases, at least put an "else".

GotoLink commented:
You call this "survival tabs" but i see no guard against creative mode. Did i miss it ?

AshIndigo commented:
I havent seen the CLA bot here, did you sign it?

This link should let you sign it: https://cla-assistant.io/MinecraftForge/MinecraftForge?pullRequest=3349

Heufneutje commented:
@AshIndigo If you click the little "Show all checks" link at the bottom of the PR you'll notice that it has indeed been signed. The bot just didn't reply for some reason.

AshIndigo commented:
@Heufneutje Oh thats where it is, thanks

AbsolemJackdaw committed:
cached container.getClass(), is now a field.
removed redundant null checks.
put a check for creative inventory. creative inventory already has tabs. trying to register a guitab for the creative container now results in nothing being registered.
shortened if/else checks to one-liners in GuiTabList
added guard against the creative inventory.
said guard gets logged as well.

diesieben07 commented:
Honestly storing the class instance like that is not worth it in my opinion. As far as I know getClass produces similar code to a normal field access, since the class object is just a field in the object header. So really this is redundant for now performance improvement.

AbsolemJackdaw commented:
I found I can reduce the container.getClass() calls from 6 to 2.
would that be sufficient ?

diesieben07 commented:
Well, you have to replace them with something, in this case a field access. The performance difference between those two should be minute, if even measurable.

TechnicianLP commented:
Do i have to instatiate the same Tab multiple times if i wanted to add itto multiple Inventories?

This API definitly has a lot of potential - but i think it is missing a few features (eg. configurable sides (think of a ae2 craftingterminal and put some tabs on both ends ....))

TechnicianLP commented on patches/minecraft/net/minecraft/client/gui/inventory/GuiContainer.java.patch:
```diff
@@ -128,11 +161,19 @@
++    	this.guiTabs.actionPerformed(button);
++    }
 +
++    public java.util.List<net.minecraftforge.client.player.inventory.tabs.GuiTab> getTabs(){
```
shouldn't this be after the ==Forge Start== ?

AbsolemJackdaw replied:
No idea. might be. I am not a forge developper or anything, so I thought i'd keep it outside for now

AbsolemJackdaw commented:
createContuity(guicontainer,guicontainer) should be the thing you want. it copies over the tabs from one inventory to another, so it creates a continuity.

if you'd like to inject one tab to more then one gui without having the need of a continuity, you can instantiate it once, and add it to the tablist when opening a gui.

i'm not sure what you are aiming for. could you add some info to that please ?

TechnicianLP commented:
Artix: continuities are a bit to much .. think of droppers and dispenser: you want a tab on both - an other mod just on the dispenser - if you created a continuity the tab from the other mod gets into unintended places ...
Having to add Tabs to the Guis manually somewhat defeats the purpose of binding them to classes

AbsolemJackdaw commented:
then indeed, creating two instances is the way to go.
I might remove that then, and make a .addTo(class<? extends GuiContainer>);
leaving the createcontinuity for the player inventory to create a continuity with gui's that have an undetermined number of tabs.

m00nwtchr commented:
cool idea shloud be merged

AbsolemJackdaw committed:
moved registry out of the constructor for a better range of registry.

refactored names to more appropriate names

adjusted code in guicontainer according to changes in guitablist and guitabs

AbsolemJackdaw committed:
list of gui tabs does not quiry by class anymore, but by a getter from the GuiTab.
allowed removal of 'createContinuity'.

AbsolemJackdaw commented:
removed continuities

mezz commented:
There are some errors with spelling in the javadocs, missing spacing, and braces on the same line.
This PR would also really benefit from including an example mod that shows the code being used.
The wiki is sort of nice but it does not fully explain things the way working example code does, and it is already outdated because it still references continuities.

mezz commented:
Can you update this to target 1.11? New features are no longer being merged into 1.10.

mezz commented:
Any update @ArtixAllMighty ?

mezz commented:
I am closing this since you have not responded in a long time.
If you decide to update this please open a new PR and link back to this one.

HenryLoenwind commented:
@ArtixAllMighty, really? I was waiting for this...

AbsolemJackdaw commented:
Native language isn't english, I make a lot of spelling errors, and the edits suggested by people are not what I had in mind for the tabs, so I have a heard time implementing them. sorry.

But AshIndigo is taking over the PR. you should see one soon that is the same, by someone else =)
